internal/graph/object_providers.go

```diff
@@ -0,0 +1,142 @@
+package graph
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/openfga/openfga/internal/checkutil"
+	"github.com/openfga/openfga/internal/concurrency"
+	openfgaErrors "github.com/openfga/openfga/internal/errors"
+	"github.com/openfga/openfga/pkg/storage"
+	"github.com/openfga/openfga/pkg/tuple"
+	"github.com/openfga/openfga/pkg/typesystem"
+)
+
+// ObjectProvider is an interface that abstracts the building of a channel that holds object IDs or usersets.
+// It must close the channel when there are no more results.
+type ObjectProvider interface {
+	End()
+	Begin(ctx context.Context, req *ResolveCheckRequest) (chan usersetMessage, error)
+}
+
+type SimpleRecursiveObjectProvider struct {
+	mapper storage.TupleMapper
+	ts     *typesystem.TypeSystem
+	ds     storage.RelationshipTupleReader
+}
+
+func NewSimpleRecursiveObjectProvider(ts *typesystem.TypeSystem, ds storage.RelationshipTupleReader) (*SimpleRecursiveObjectProvider, error) {
+	if ts == nil || ds == nil {
+		return nil, fmt.Errorf("%w: nil arguments", openfgaErrors.ErrUnknown)
+	}
+	return &SimpleRecursiveObjectProvider{ts: ts, ds: ds}, nil
+}
+
+var _ ObjectProvider = (*SimpleRecursiveObjectProvider)(nil)
+
+func (s *SimpleRecursiveObjectProvider) End() {
+	if s.mapper != nil {
+		s.mapper.Stop()
+	}
+}
+
+//nolint:revive // TODO make usersetMessage an exported struct
+func (s *SimpleRecursiveObjectProvider) Begin(ctx context.Context, req *ResolveCheckRequest) (chan usersetMessage, error) {
+	if req == nil {
+		return nil, fmt.Errorf("%w: nil request", openfgaErrors.ErrUnknown)
+	}
+	// Note: we set sortContextualTuples to false because we don't care about ordering of results,
+	// since the consumer is using hashsets to check for intersection.
+	userIter, err := checkutil.IteratorReadStartingFromUser(ctx, s.ts, s.ds, req,
+		tuple.ToObjectRelationString(tuple.GetType(req.GetTupleKey().GetObject()), req.GetTupleKey().GetRelation()),
+		nil, false)
+	if err != nil {
+		return nil, err
+	}
+	usersetFromUserIter := storage.WrapIterator(storage.ObjectIDKind, userIter)
+	s.mapper = usersetFromUserIter
+
+	// note: this function will close the channel
+	userToUsersetMessageChan := streamedLookupUsersetFromIterator(ctx, usersetFromUserIter)
+
+	return userToUsersetMessageChan, nil
+}
+
+type ComplexRecursiveObjectProvider struct {
+	concurrencyLimit uint32
+	wg               sync.WaitGroup
+	ts               *typesystem.TypeSystem
+}
+
+func NewComplexRecursiveObjectProvider(concurrencyLimit uint32, ts *typesystem.TypeSystem) (*ComplexRecursiveObjectProvider, error) {
+	if ts == nil {
+		return nil, fmt.Errorf("%w: nil typesystem", openfgaErrors.ErrUnknown)
+	}
+	return &ComplexRecursiveObjectProvider{concurrencyLimit: concurrencyLimit, wg: sync.WaitGroup{}, ts: ts}, nil
```

Comment:
Undefined limit would be 0, would that panic or be unbounded?

Reply:
If the limit were 0, it would panic here: 
<img width="638" alt="image" src="https://github.com/user-attachments/assets/cc6c7089-5193-4013-9235-db9a83a14df2" />

I can send a different PR to ensure that doesn't happen